File: ios/Battleships/UI/Screens/BattleScreen.swift
import SwiftUI

struct BattleScreen: View {
    @ObservedObject var vm: GameViewModel

    private var s: I18nStrings { vm.s }

    var body: some View {
        ScrollView {
            VStack(spacing: 16) {
                // Timer
                if !vm.playerTimeLeft.isEmpty {
                    GameTimerView(
                        playerTimeLeft: vm.playerTimeLeft,
                        turnStartedAt: vm.turnStartedAt,
                        currentTurn: vm.currentTurn,
                        myId: vm.playerIdRef,
                        opponentName: vm.opponentName,
                        strings: s,
                        spectatorPlayerNames: vm.isSpectator
                            ? Dictionary(uniqueKeysWithValues: vm.spectatorBoards.map { ($0.playerId, $0.playerName) })
                            : [:]
                    )
                }

                // Turn indicator
                turnIndicator

                // Scoreboard (non-spectator)
                if !vm.isSpectator {
                    scoreBoard
                }

                // Spectator boards or normal boards
                if vm.isSpectator {
                    spectatorBoards
                } else {
                    normalBoards
                }
            }
            .padding(.horizontal, 12)
            .padding(.top, 8)
            .padding(.bottom, 40)
        }
        .alert(s.confirmSurrender, isPresented: $vm.showSurrenderDialog) {
            Button(s.yes, role: .destructive) { vm.confirmForfeit() }
            Button(s.no, role: .cancel) { vm.cancelForfeit() }
        }
    }

    // MARK: - Turn Indicator
    private var turnIndicator: some View {
        HStack {
            VStack(alignment: .leading, spacing: 2) {
                HStack(spacing: 6) {
                    if vm.isSpectator {
                        Text("👁️").foregroundColor(.purple)
                        Text(s.spectating).font(.headline).foregroundColor(.purple)
                        if let turnName = vm.spectatorBoards.first(where: { $0.playerId == vm.currentTurn })?.playerName {
                            Text("— \(s.namesTurn.fmt(turnName))")
                                .font(.caption)
                                .foregroundColor(.gray)
                        }
                    } else if vm.isMyTurn {
                        Text("⚡").foregroundColor(.green)
                        Text(s.yourTurnFire).font(.headline).foregroundColor(.green)
                    } else {
                        Text("⏳").foregroundColor(.gray)
                        Text(s.namesTurn.fmt(vm.opponentName)).font(.headline).foregroundColor(.white)
                    }
                }

                Text(!vm.isSpectator && vm.isMyTurn ? s.extraShotHint : " ")
                    .font(.caption2)
                    .foregroundColor(.green.opacity(vm.isMyTurn ? 0.6 : 0))
            }

            Spacer()

            HStack(spacing: 8) {
                if vm.spectatorCount > 0 {
                    Text("👁️ \(vm.spectatorCount)")
                        .font(.caption2.bold())
                        .padding(.horizontal, 8)
                        .padding(.vertical, 6)
                        .background(Color.purple.opacity(0.2))
                        .cornerRadius(8)
                        .foregroundColor(.purple)
                }

                Button {
                    if vm.isSpectator {
                        vm.handleBackToMenu()
                    } else {
                        vm.requestForfeit()
                    }
                } label: {
                    Text(vm.isSpectator ? s.leave : "🏳️")
                        .font(.caption.bold())
                        .padding(.horizontal, 10)
                        .padding(.vertical, 8)
                        .overlay(RoundedRectangle(cornerRadius: 8).stroke(Color.red.opacity(0.3)))
                        .foregroundColor(.red.opacity(0.8))
                        .cornerRadius(8)
                }
            }
        }
        .padding(14)
        .background(
            RoundedRectangle(cornerRadius: 14)
                .fill(Color.white.opacity(0.06))
                .overlay(RoundedRectangle(cornerRadius: 14).stroke(
                    vm.isSpectator
                        ? Color.purple.opacity(0.3)
                        : vm.isMyTurn ? Color.green.opacity(0.3) : Color.white.opacity(0.08),
                    lineWidth: 1))
        )
    }

    // MARK: - Score Board
    private var scoreBoard: some View {
        HStack(spacing: 16) {
            HStack(spacing: 6) {
                Text(s.yourHits).font(.caption.bold()).foregroundColor(.green)
                HStack(spacing: 3) {
                    ForEach(0..<5, id: \.self) { i in
                        Circle()
                            .fill(i < vm.theirSunkCount ? Color.orange : Color.white.opacity(0.1))
                            .frame(width: 8, height: 8)
                    }
                }
            }
            Text("│").foregroundColor(.gray.opacity(0.3))
            HStack(spacing: 6) {
                HStack(spacing: 3) {
                    ForEach(0..<5, id: \.self) { i in
                        Circle()
                            .fill(i < vm.mySunkCount ? Color.red : Color.white.opacity(0.1))
                            .frame(width: 8, height: 8)
                    }
                }
                Text(s.theirHits).font(.caption.bold()).foregroundColor(.red)
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 10)
        .background(
            RoundedRectangle(cornerRadius: 12)
                .fill(Color.white.opacity(0.04))
                .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color.white.opacity(0.06)))
        )
    }

    // MARK: - Normal Boards
    private var normalBoards: some View {
        VStack(spacing: 14) {
            // Your fleet
            boardCard(label: s.yourFleet, isHighlighted: false) {
                GameBoardView(
                    board: vm.playerBoard,
                    isOpponentBoard: false,
                    isInteractive: false,
                    shotKeys: vm.playerShotKeys,
                    explosionKeys: vm.playerExplosionKeys
                )
            }

            // Enemy waters
            boardCard(label: s.enemyWaters.fmt(vm.opponentName), isHighlighted: vm.isMyTurn) {
                GameBoardView(
                    board: vm.opponentBoard,
                    isOpponentBoard: true,
                    isInteractive: vm.isMyTurn,
                    shotKeys: vm.shotKeys,
                    explosionKeys: vm.explosionKeys
                ) { row, col in
                    vm.shoot(row: row, col: col)
                }
            }
        }
    }

    // MARK: - Spectator Boards
    private var spectatorBoards: some View {
        VStack(spacing: 14) {
            ForEach(vm.spectatorBoards, id: \.playerId) { sb in
                boardCard(label: "🛡️ \(sb.playerName)", isHighlighted: false) {
                    GameBoardView(
                        board: sb.board,
                        isOpponentBoard: false,
                        isInteractive: false,
                        shotKeys: [],
                        explosionKeys: []
                    )
                }
            }
        }
    }

    @ViewBuilder
    private func boardCard<Content: View>(label: String, isHighlighted: Bool, @ViewBuilder content: () -> Content) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(label)
                .font(.subheadline.bold())
                .foregroundColor(.white)

            content()

            // Legend
            HStack(spacing: 10) {
                legendItem(.blue.opacity(0.7), "Water")
                legendItem(.green.opacity(0.6), "Ship")
                legendItem(.red.opacity(0.8), "Hit")
                legendItem(.gray.opacity(0.5), "Miss")
                legendItem(Color(red: 0.6, green: 0.1, blue: 0.1), "Sunk")
            }
            .font(.system(size: 9))
            .foregroundColor(.gray)
        }
        .padding(12)
        .background(
            RoundedRectangle(cornerRadius: 14)
                .fill(Color.white.opacity(0.04))
                .overlay(RoundedRectangle(cornerRadius: 14).stroke(
                    isHighlighted ? Color.green.opacity(0.2) : Color.white.opacity(0.06), lineWidth: 1))
        )
    }

    private func legendItem(_ color: Color, _ label: String) -> some View {
        HStack(spacing: 3) {
            RoundedRectangle(cornerRadius: 2).fill(color).frame(width: 8, height: 8)
            Text(label)
        }
    }
}

File: android/app/src/main/java/com/anasio/battleships/ui/screens/BattleScreen.kt
package com.anasio.battleships.ui.screens

import androidx.compose.animation.*
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.sp
import com.anasio.battleships.data.CellState
import com.anasio.battleships.i18n.LocalI18n
import com.anasio.battleships.i18n.fmt
import com.anasio.battleships.ui.components.GameBoard
import com.anasio.battleships.ui.components.GameTimer
import com.anasio.battleships.ui.theme.*
import com.anasio.battleships.ui.theme.LocalColorPalette
import com.anasio.battleships.viewmodel.GameViewModel

@Composable
fun BattleScreen(viewModel: GameViewModel) {
    val playerBoard by viewModel.playerBoard.collectAsState()
    val opponentBoard by viewModel.opponentBoard.collectAsState()
    val currentTurn by viewModel.currentTurn.collectAsState()
    val myId by viewModel.playerId.collectAsState()
    val opponentName by viewModel.opponentName.collectAsState()
    val message by viewModel.message.collectAsState()
    val messageType by viewModel.messageType.collectAsState()
    val isSpectator by viewModel.isSpectator.collectAsState()
    val spectatorCount by viewModel.spectatorCount.collectAsState()
    val spectatorBoards by viewModel.spectatorBoards.collectAsState()
    val playerTimeLeft by viewModel.playerTimeLeft.collectAsState()
    val turnStartedAt by viewModel.turnStartedAt.collectAsState()
    val soundEnabled by viewModel.soundEnabled.collectAsState()
    val musicEnabled by viewModel.musicEnabled.collectAsState()
    val showSurrenderDialog by viewModel.showSurrenderDialog.collectAsState()
    val s = LocalI18n.current
    val c = LocalColorPalette.current

    val mySunkCount by viewModel.mySunkCount.collectAsState()
    val theirSunkCount by viewModel.theirSunkCount.collectAsState()
    val lastShotKey by viewModel.lastShotKey.collectAsState()
    val explosionKeys by viewModel.explosionKeys.collectAsState()

    val isMyTurn by remember { derivedStateOf { !isSpectator && currentTurn == myId } }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = 8.dp, vertical = 4.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        // Audio toggles
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.End,
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(
                onClick = viewModel::toggleSound,
                modifier = Modifier.size(44.dp).semantics { contentDescription = s.toggleSound },
            ) {
                Text(if (soundEnabled) "🔊" else "🔇", fontSize = 18.sp)
            }
            IconButton(
                onClick = viewModel::toggleMusic,
                modifier = Modifier.size(44.dp).semantics { contentDescription = s.toggleMusic },
            ) {
                Text(if (musicEnabled) "🎵" else "🔕", fontSize = 18.sp)
            }
        }

        // Timer
        val spectatorNameMap = remember(spectatorBoards) {
            spectatorBoards.associateBy({ it.playerId }, { it.playerName })
        }
        if (playerTimeLeft.isNotEmpty()) {
            GameTimer(
                playerTimeLeft = playerTimeLeft,
                turnStartedAt = turnStartedAt,
                currentTurn = currentTurn,
                myId = myId,
                opponentName = opponentName,
                isSpectator = isSpectator,
                spectatorPlayerNames = spectatorNameMap,
            )
            Spacer(Modifier.height(4.dp))
        }

        // Turn indicator
        val turnText = when {
            isSpectator -> {
                val name = spectatorBoards.find { it.playerId == currentTurn }?.playerName ?: "?"
                s.namesTurn.fmt(name)
            }
            isMyTurn -> "🎯 ${s.yourTurnFire}"
            else -> "⏳ ${s.opponentsTurn}"
        }
        val turnColor = when {
            isSpectator -> c.primary
            isMyTurn -> c.green
            else -> c.orange
        }
        AnimatedContent(
            targetState = isMyTurn,
            transitionSpec = {
                (fadeIn() + slideInVertically { -it / 2 }) togetherWith
                        (fadeOut() + slideOutVertically { it / 2 })
            },
            label = "turnTransition",
        ) { myTurn ->
            val text = if (myTurn) "🎯 ${s.yourTurnFire}" else if (isSpectator) turnText else "⏳ ${s.opponentsTurn}"
            val color = if (myTurn) c.green else if (isSpectator) c.primary else c.orange
            Text(
                text,
                fontSize = 15.sp, fontWeight = FontWeight.Bold, color = color,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .fillMaxWidth()
                    .background(color.copy(alpha = .1f), RoundedCornerShape(8.dp))
                    .border(1.dp, color.copy(alpha = .3f), RoundedCornerShape(8.dp))
                    .padding(vertical = 6.dp),
            )
        }
        // Extra shot hint (always rendered to prevent layout shift)
        Text(
            s.extraShotHint, fontSize = 10.sp, color = c.textDim,
            textAlign = TextAlign.Center,
            modifier = Modifier.fillMaxWidth().alpha(if (isMyTurn) 1f else 0f),
        )
        Spacer(Modifier.height(4.dp))

        // Ship scoreboard
        if (!isSpectator) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceEvenly,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                SunkDots(label = s.yourHits, count = mySunkCount, dotColor = c.green)
                SunkDots(label = s.theirHits, count = theirSunkCount, dotColor = c.red)
            }
            Spacer(Modifier.height(4.dp))
        }

        // Message
        if (message.isNotBlank()) {
            MessageBanner(message, messageType)
            Spacer(Modifier.height(2.dp))
        }

        if (isSpectator) {
            // Spectator: show both boards
            spectatorBoards.forEach { sb ->
                GameBoard(
                    board = sb.board,
                    label = sb.playerName,
                    showShips = true,
                    interactive = false,
                )
                Spacer(Modifier.height(8.dp))
            }
        } else {
            // Enemy waters (interactive)
            GameBoard(
                board = opponentBoard,
                label = "🎯 ${s.enemyWaters}",
                showShips = false,
                interactive = isMyTurn,
                lastShotKey = lastShotKey,
                explosionKeys = explosionKeys,
                onCellClick = { r, col ->
                    val cell = opponentBoard[r][col]
                    if (cell == CellState.WATER) viewModel.handleShoot(r, col)
                },
            )
            Spacer(Modifier.height(8.dp))
            // Your fleet (read only)
            GameBoard(
                board = playerBoard,
                label = "🚥 ${s.yourFleet}",
                showShips = true,
                interactive = false,
            )
        }

        Spacer(Modifier.height(8.dp))

        // Footer
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (spectatorCount > 0) {
                Text("👁 $spectatorCount ${s.spectatorCount}", fontSize = 11.sp, color = c.textDim)
            } else Spacer(Modifier.width(1.dp))

            if (!isSpectator) {
                OutlinedButton(
                    onClick = viewModel::requestForfeit,
                    colors = ButtonDefaults.outlinedButtonColors(contentColor = c.red),
                    modifier = Modifier.height(44.dp),
                    shape = RoundedCornerShape(8.dp),
                ) { Text("🏳 ${s.surrender}", fontSize = 12.sp) }
            } else {
                OutlinedButton(
                    onClick = viewModel::handleBackToMenu,
                    modifier = Modifier.height(44.dp),
                    shape = RoundedCornerShape(8.dp),
                ) { Text("← ${s.leave}", fontSize = 12.sp) }
            }
        }
        Spacer(Modifier.height(12.dp))
    }

    if (showSurrenderDialog) {
        AlertDialog(
            onDismissRequest = viewModel::cancelForfeit,
            title = { Text(s.surrender) },
            text = { Text(s.confirmSurrender) },
            confirmButton = {
                TextButton(onClick = viewModel::confirmForfeit) { Text(s.yes) }
            },
            dismissButton = {
                TextButton(onClick = viewModel::cancelForfeit) { Text(s.no) }
            },
        )
    }
}

@Composable
private fun SunkDots(label: String, count: Int, dotColor: Color) {
    val c = LocalColorPalette.current
    Row(verticalAlignment = Alignment.CenterVertically) {
        Text(label, fontSize = 10.sp, color = dotColor, fontWeight = FontWeight.Bold)
        Spacer(Modifier.width(4.dp))
        repeat(5) { i ->
            Box(
                modifier = Modifier
                    .size(10.dp)
                    .padding(1.dp)
                    .clip(RoundedCornerShape(50))
                    .background(if (i < count) dotColor else c.card)
                    .border(0.5.dp, if (i < count) dotColor else c.border, RoundedCornerShape(50)),
            )
        }
    }
}
